recipes-core/searcher-ssh-key/files/mode-controller.c

```diff
@@ -0,0 +1,391 @@
+#define _GNU_SOURCE
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+#include <time.h>
+#include <sys/file.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <signal.h>
+#include <fcntl.h>
+#include <syslog.h>
+#include <stdarg.h>
+
+#define STATE_FILE "/run/system-mode.state"
+#define LOCK_FILE "/run/mode-transition.lock"
+#define MODE_MAINTENANCE "maintenance"
+#define MODE_PRODUCTION "production"
+#define MODE_TRANSITION "transition"
+#define SEARCHER_UID 1000
+#define TIMEOUT_SECONDS 300  // 5 minutes
+
+static int lock_fd = -1;
+
+// Function prototypes
+static int acquire_lock(void);
+static void release_lock(void);
+static int read_current_mode(char *mode, size_t size);
+static int write_mode(const char *mode);
+static int execute_transition_script(const char *from_mode, const char *to_mode);
+static void log_message(int priority, const char *format, ...);
+static int kill_existing_connections(void);
+
+static void log_message(int priority, const char *format, ...) {
+    va_list args;
+    va_start(args, format);
+    vsyslog(priority, format, args);
+    vfprintf(stderr, format, args);
+    fprintf(stderr, "\n");
+    va_end(args);
+}
+
+static int acquire_lock(void) {
+    if (lock_fd != -1) return 0;  // Already have lock
+
+    lock_fd = open(LOCK_FILE, O_RDWR | O_CREAT, 0600);
+    if (lock_fd == -1) {
+        log_message(LOG_ERR, "Failed to open lock file: %s", strerror(errno));
+        return -1;
+    }
+
+    if (flock(lock_fd, LOCK_EX | LOCK_NB) == -1) {
+        close(lock_fd);
+        lock_fd = -1;
+        log_message(LOG_ERR, "Failed to acquire lock: %s", strerror(errno));
+        return -1;
+    }
+    return 0;
+}
+
+static void release_lock(void) {
+    if (lock_fd != -1) {
+        flock(lock_fd, LOCK_UN);
+        close(lock_fd);
+        lock_fd = -1;
+    }
+}
+
+static int read_current_mode(char *mode, size_t size) {
+    if (acquire_lock() != 0) {
```

Comment:
I'd probably lock the whole execution -- I don't think concurrent access is handled

Reply:
So you mean not on function level but rather directly in the main. Lock it and only unlock it when finished instead?